src/ol/renderer/canvas/TileLayer.js

```diff
@@ -329,22 +329,22 @@ class CanvasTileLayerRenderer extends CanvasLayerRenderer {
         continue;
       }
 
+      const gutter = Math.round(
+        tilePixelRatio * source.getGutterForProjection(viewState.projection),
```

Comment:
Here you're starting to use the `tilePixelRatio`, below you're getting rid of it. I don't think this will work when a gutter is actually specified. At least we should be clear about whether the gutter is meant to be provided in css pixels or canvas pixels.

Reply:
As noted in #16467 gutter does need to be multiplied by the pixel ratio (either 1 or the map pixel ratio) for TileWMS.  In other cases it should not, but canvas and WebGL are currently inconsistent in places.  If other references to `tilePixelRatio` are removed (e.g. canvas tile reprojection) `source.getTilePixelRatio()` could return 1 for everything except TileWMS (and perhaps Zoomify and IIIF if necessary - they do not have a gutter).

Reply:
The docs say "The size in pixels", which is not very specific. If your plan is to get rid of having to specify a `tilePixelRatio`, then `gutter` needs to be in *image* pixels.

Reply:
That is how is has to work with `ol/source/ImageTile` and `DataTile` and it makes sense for `XYZ` to match that. `TileWMS` is different because it buffers the extent, using the same extent regardless of whether `hidpi` or not (which also makes sense as any difference could affect label positions, etc.) hence the need to multiply gutter here, but that is internal as it does not have a `tilePixelRatio` option, and the ratio used depends on the map pixel ratio.

Reply:
However, this means that the `gutter` option would have a different meaning for TileWMS. I'd be in favor of clarifying once and for all that gutter is in image pixels, and I don't see why TileWMS should be an exception.

Reply:
TileWMS always has been different, the gutter option is used in a request to create an image which is also uses the map pixel ratio, and that can differ when the same application is used on different devices.  A fixed bbox is needed to avoid (for example) label positioning differing on those devices.  Also the original purpose was to avoid tile edge issues originating from the server.  In XYZ and ImageTile source tiles from sources such as Nextzen and Mapbox have predefined fixed gutters and pixel ratios, and the purpose is to allow interpolation and shading calculations across tile edges, where a one pixel gutter will do the same job regardless of any pixel ratio.  I agree we should not be calling `getTilePixelRatio()` but from a rendering perspective the image pixels can be used everywhere if `getGutterForProjection()` considers the map pixel ratio as well as projection where necessary.